File: python/experiments/asymmetric/variable/split_depth.py
import math
import time
import contextlib
import io
import csv
import os

from asymmetric_sequential import generate_fractal_tree_asymmetric
from asymmetric_parallel import run_parallel_asymmetric

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', '..', '..'))
OUTPUT_DIR   = os.path.join(PROJECT_ROOT, 'data', 'asymmetric', 'split_depth')
CSV_PATH     = os.path.join(OUTPUT_DIR, 'empirical_python.csv')

# Variable split_depth experiment — asymmetric tree, Python
# Fixed: left_ratio=0.67, right_ratio=0.57, min_length=0.01, cores=8
#
# split_depth varies: 1 to 12
# Each depth is run NUM_RUNS times; mean is reported.
#
# Purpose: find the empirically optimal split_depth for Python multiprocessing.
# Hypothesis: deeper splits increase pickle/IPC overhead and worsen performance
# even though T_worst analysis predicts improvement at depth ~11.

NUM_PROCESSES = 8
NUM_RUNS      = 3
TRUNK_LENGTH  = 100.0
LEFT_RATIO    = 0.67
RIGHT_RATIO   = 0.57
LEFT_ANGLE    = 35.0
RIGHT_ANGLE   = 25.0
MIN_LENGTH    = 0.01
SPLIT_DEPTHS  = list(range(1, 13))


def _time_sequential():
    left_rad  = math.radians(LEFT_ANGLE)
    right_rad = math.radians(RIGHT_ANGLE)
    t0 = time.perf_counter()
    branches = generate_fractal_tree_asymmetric(
        0, 0, TRUNK_LENGTH, math.pi / 2,
        LEFT_RATIO, RIGHT_RATIO, left_rad, right_rad, MIN_LENGTH,
    )
    return time.perf_counter() - t0, len(branches)


def _time_parallel(split_depth):
    with contextlib.redirect_stdout(io.StringIO()):
        result = run_parallel_asymmetric(
            trunk_length=TRUNK_LENGTH,
            left_ratio=LEFT_RATIO,
            right_ratio=RIGHT_RATIO,
            left_angle=LEFT_ANGLE,
            right_angle=RIGHT_ANGLE,
            min_length=MIN_LENGTH,
            num_processes=NUM_PROCESSES,
            split_depth=split_depth,
        )
    return result['execution_time']


if __name__ == '__main__':
    seq_time, branch_count = _time_sequential()
    num_tasks_at = lambda d: 2 ** d

    print(f"\n=== Variable split_depth (asymmetric, Python) | cores={NUM_PROCESSES}, "
          f"left_ratio={LEFT_RATIO}, right_ratio={RIGHT_RATIO}, min_length={MIN_LENGTH} ===")
    print(f"  Sequential baseline: {seq_time:.5f}s  ({branch_count:,} branches)")
    print()
    print(f"{'depth':>6} {'tasks':>7} {'par_mean (s)':>13} {'speedup':>9} {'efficiency':>12}  note")
    print("-" * 65)

    rows = []
    current_heuristic = max(1, math.ceil(math.log2(NUM_PROCESSES * 4)))

    for depth in SPLIT_DEPTHS:
        times = [_time_parallel(depth) for _ in range(NUM_RUNS)]
        par_mean  = sum(times) / NUM_RUNS
        speedup   = seq_time / par_mean
        efficiency = speedup / NUM_PROCESSES
        note = "<-- heuristic" if depth == current_heuristic else ""

        print(f"{depth:>6} {num_tasks_at(depth):>7,} {par_mean:>13.5f} {speedup:>8.3f}x {efficiency:>11.1%}  {note}")
        rows.append({
            'split_depth': depth,
            'num_tasks':   num_tasks_at(depth),
            'branches':    branch_count,
            'seq_time':    f'{seq_time:.6f}',
            'par_mean':    f'{par_mean:.6f}',
            'speedup':     f'{speedup:.4f}',
            'efficiency':  f'{efficiency:.4f}',
        })

    os.makedirs(OUTPUT_DIR, exist_ok=True)
    with open(CSV_PATH, 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=[
            'split_depth', 'num_tasks', 'branches', 'seq_time', 'par_mean', 'speedup', 'efficiency'
        ])
        writer.writeheader()
        writer.writerows(rows)
    print(f"\nSaved: {CSV_PATH}")

File: rust/experiments/asymmetric/split_depth.rs
use fractal_tree::{generate_fractal_tree_asymmetric, collect_tasks_asymmetric, Branch};
use rayon::prelude::*;
use std::f64::consts::PI;
use std::time::Instant;
use std::fs;
use std::io::Write;
use std::path::Path;

const NUM_THREADS: usize = 8;
const NUM_RUNS:    usize = 3;
const MIN_LENGTH:  f64   = 0.01;
const TRUNK:       f64   = 100.0;
const LEFT_RATIO:  f64   = 0.67;
const RIGHT_RATIO: f64   = 0.57;
const LEFT_ANGLE:  f64   = 35.0;   // degrees
const RIGHT_ANGLE: f64   = 25.0;   // degrees
const MAX_DEPTH:   usize = 12;

// Variable split_depth experiment — asymmetric tree, Rust/Rayon
// Fixed: left_ratio=0.67, right_ratio=0.57, min_length=0.01, threads=8
//
// split_depth varies: 1 to 12
// Each depth is run NUM_RUNS times; mean is reported.
//
// Purpose: find the empirically optimal split_depth for Rayon work-stealing.
// Hypothesis: shallow splits are sufficient because work-stealing handles
// load imbalance dynamically — unlike Python's static pool.map.

fn run_parallel(left_angle_rad: f64, right_angle_rad: f64, split: usize, pool: &rayon::ThreadPool) -> f64 {
    let start = Instant::now();

    let mut upper = vec![Branch { x1: 0.0, y1: 0.0, x2: 0.0, y2: TRUNK, depth: 0 }];
    let mut tasks = collect_tasks_asymmetric(
        &mut upper,
        0.0, TRUNK, TRUNK * LEFT_RATIO, PI / 2.0 + left_angle_rad,
        LEFT_RATIO, RIGHT_RATIO, left_angle_rad, right_angle_rad, 1, split,
    );
    tasks.extend(collect_tasks_asymmetric(
        &mut upper,
        0.0, TRUNK, TRUNK * RIGHT_RATIO, PI / 2.0 - right_angle_rad,
        LEFT_RATIO, RIGHT_RATIO, left_angle_rad, right_angle_rad, 1, split,
    ));

    let _results: Vec<Vec<Branch>> = pool.install(|| {
        tasks.par_iter().map(|t| generate_fractal_tree_asymmetric(
            t.x, t.y, t.length, t.angle,
            LEFT_RATIO, RIGHT_RATIO, left_angle_rad, right_angle_rad, MIN_LENGTH, t.depth,
        )).collect()
    });

    start.elapsed().as_secs_f64()
}

fn main() {
    let left_angle_rad  = LEFT_ANGLE.to_radians();
    let right_angle_rad = RIGHT_ANGLE.to_radians();
    let pool = rayon::ThreadPoolBuilder::new().num_threads(NUM_THREADS).build().unwrap();
    let heuristic_depth = (NUM_THREADS * 4).next_power_of_two().ilog2() as usize;

    // sequential baseline
    let seq_start = Instant::now();
    let seq_branches = generate_fractal_tree_asymmetric(
        0.0, 0.0, TRUNK, PI / 2.0,
        LEFT_RATIO, RIGHT_RATIO, left_angle_rad, right_angle_rad, MIN_LENGTH, 0,
    );
    let seq_time = seq_start.elapsed().as_secs_f64();
    let branch_count = seq_branches.len();

    println!("\n=== Variable split_depth (asymmetric, Rust) | threads={NUM_THREADS}, \
              left_ratio={LEFT_RATIO}, right_ratio={RIGHT_RATIO}, min_length={MIN_LENGTH} ===");
    println!("  Sequential baseline: {seq_time:.5}s  ({branch_count} branches)");
    println!();
    println!("{:>6} {:>7} {:>13} {:>9} {:>12}  note",
             "depth", "tasks", "par_mean (s)", "speedup", "efficiency");
    println!("{}", "-".repeat(65));

    let mut csv_rows: Vec<String> = Vec::new();
    csv_rows.push("split_depth,num_tasks,branches,seq_time,par_mean,speedup,efficiency".to_string());

    for depth in 1..=MAX_DEPTH {
        let num_tasks = 2_usize.pow(depth as u32);
        let times: Vec<f64> = (0..NUM_RUNS)
            .map(|_| run_parallel(left_angle_rad, right_angle_rad, depth, &pool))
            .collect();
        let par_mean   = times.iter().sum::<f64>() / NUM_RUNS as f64;
        let speedup    = seq_time / par_mean;
        let efficiency = speedup / NUM_THREADS as f64;
        let note = if depth == heuristic_depth { "<-- heuristic" } else { "" };

        println!("{depth:>6} {num_tasks:>7} {par_mean:>13.5} {speedup:>8.3}x {pct:>11.1}%  {note}",
                 pct = efficiency * 100.0);

        csv_rows.push(format!("{},{},{},{:.6},{:.6},{:.4},{:.4}",
                              depth, num_tasks, branch_count, seq_time, par_mean, speedup, efficiency));
    }

    // save CSV
    let out_dir = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent().unwrap()
        .join("data").join("asymmetric").join("split_depth");
    fs::create_dir_all(&out_dir).unwrap();
    let csv_path = out_dir.join("empirical_rust.csv");
    let mut file = fs::File::create(&csv_path).unwrap();
    for row in &csv_rows {
        writeln!(file, "{row}").unwrap();
    }
    println!("\nSaved: {}", csv_path.display());
}
